Dockerfile

```diff
@@ -1,21 +1,45 @@
-ARG NODE_ENV=production
-ARG NODE_VERSION=10.15.3
-
-# Production
-FROM node:${NODE_VERSION}-alpine AS production
-
-ARG NODE_ENV
-
-ENV NODE_ENV $NODE_ENV
-
-USER node
-WORKDIR /home/node
-
-EXPOSE 3001
-
-COPY --chown=node:node package.json package-lock.json /home/node/
-COPY --chown=node:node server /home/node/server/
-COPY --chown=node:node index.js /home/node/index.js
-RUN npm ci
-
-CMD ["node", "index.js"]
+ARG NODE_VERSION=10.15.3
+
+# Production base
+FROM node:${NODE_VERSION}-alpine AS prodbase
+
+USER node
+WORKDIR /home/node
+
+ENV NODE_ENV production
+
+COPY --chown=node:node package.json package-lock.json /home/node/
+RUN npm ci --loglevel verbose
+
+# Development
+FROM node:${NODE_VERSION}-alpine AS development
+
+USER node
+WORKDIR /home/node
+
+ENV NODE_ENV development
+
+COPY --chown=node:node package.json package-lock.json /home/node/
+RUN npm ci --loglevel verbose
+
+COPY --chown=node:node . /home/node/
+
+CMD ["node", "index.js"]
+
+# Production
+FROM node:${NODE_VERSION}-alpine AS production
+
+ARG NODE_ENV=production
+ENV NODE_ENV $NODE_ENV
+
+USER node
+WORKDIR /home/node
+
+EXPOSE 3001
+
+COPY --chown=node:node --from=prodbase /home/node/package.json /home/node/package-lock.json /home/node/
+COPY --chown=node:node --from=prodbase /home/node/node_modules /home/node/node_modules/
+COPY --chown=node:node --from=development /home/node/server /home/node/server/
+COPY --chown=node:node --from=development /home/node/index.js /home/node/index.js
+
+CMD ["node", "index.js"]
```

Comment:
Good point, I think this got copied across from the web project, where we do a SASS build in the dev stage and copy the result to production. We don't compile/transpile any code here so there's really no point in having a dev stage at all.

Reply:
I've restructured the Dockerfile. I've kept two stages so that the production image doesn't include the test folder. however, I've put the production stage first with the development stage extending the production image to add in tests and dev dependencies.

Reply:
I've now also realised that copying our code in before the development stage means the dev `npm install` is re-run on every container rebuild with a code change. I've rearranged again so that we now do a production `npm install` in a base stage, which is extended by each of the development and production stages. This arrangement means most Docker builds in development should use cached layers for everything except copying in code changes.